syscalls/fsetxattr: Add fsetxattr() tests based on setxattr() ones

rafaeldtinoco opened the pull request:
Fixes: #274

Following the same logic and tests used to test setxattr() syscalls,
this commit implements tests for fsetxattr(). It only differs from
setxattr() on the given arguments: using a file descriptor instead of
the filename.

Kernel has different entry points for both, with slightly different
execution paths, mainly related to dealing with the passed file
descriptor. Minor changes were made in order for the file descriptors
to be available for the fsetxattr() call.

Signed-off-by: Rafael David Tinoco <rafael.tinoco@linaro.org>

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,231 @@
+		SAFE_BIND(fd, (const struct sockaddr *) &sun,
+				sizeof(struct sockaddr_un));
+	}
+
```
Hardcoding offsets to the table describing testcases is usually frowned upon.

Maybe we can replace the filename in test structure with filedescriptor and initialize all these in the setup.

Something as:
```
static int file_fd;
static int dir_fd;
...
static struct test_case {
        int *fd;
        ...
} tc[] = {
       {.fd = &fd_file, ...},
       ...
};
...
static void setup(void)
{
       fd_file = SAFE_OPEN(...);
       ...
}
```

rafaeldtinoco replied:
Yep. I had doubts also. Will change it. Tks

rafaeldtinoco replied:
Made the changes, let me know what you think, please. Cheers!

rafaeldtinoco commented:
CI build timed out in one of the builders (apt-get) and I pushed same commit again to see if all builders pass this time.

pevik commented:
@rafaeldtinoco I restarted it for you. Unfortunately it happens sometimes.

rafaeldtinoco commented:
Hello @pevik and @metan-ucw, should I mail this to the list ? or merge requests through here are equally fine ? I'm waiting on acceptance before moving on in other syscall tests, just to make sure its all good on my side (not to replicate wrong assumptions over and over). Cheers!

metan-ucw commented:
I will try to have a look ASAP, the problem is that there will be public holiday in Czech Republic for two days then we will have hackweek in SUSE, so there is very little room left for review at least from SUSE until Monday 16.

metan-ucw commented:
Also I do prefer emails to mailing list a bit since that is what I'm more used to, but I do work with pull request here as well.

rafaeldtinoco commented:
Alright. Will keep this one like this then, possibly adding fgetxattr() tests to it, and the next one will send over the mailing list. Tks!

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,267 @@
+		TP_CHARDEV,
+		TP_BLOCKDEV,
+		TP_SOCKET
+	} type;
```
We do use this enum only to distinguish if we are supposed to create socket in the test setup, right?

Shouldn't a issocket flag suffice here?

rafaeldtinoco replied:
Yes, I thought about being extensible for future needs to be honest. I'll change it and, if ever needed, we can use it later.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,267 @@
+	{			/* case 00, set attr to reg */
+	 .type = TP_FILE,
+	 .fname = FILENAME,
+	 .filed = -1,
```
I guess that there is no point in setting the file descriptor to -1 here, since we check for > 0  in the test cleanup we can as well have it auto-initialized to 0.

Also fd is probably a bit more common name for a variable that represents file descriptor, but that is very minor...

rafaeldtinoco replied:
Agreed. Since I'll change this I'll rename the variable as well.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,267 @@
+	}
+
+	tst_res(TPASS | TTERRNO, "fsetxattr() failed");
+}
```
It would probably be a bit nicer if we included what type of fd we were testing, i.e. use the fname in the tst_res() messages so that we have something as:

```
fsetxattr02.c:87: PASS: fsetxattr() on testdir passed
fsetxattr02.c:87: PASS: fsetxattr() on symlink failed: EEXIST
````

rather than:

```
fsetxattr02.c:87: PASS: fsetxattr() passed
fsetxattr02.c:87: PASS: fsetxattr() failed: EEXIST
```

rafaeldtinoco replied:
Will do.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,202 @@
+	 .flags = XATTR_CREATE,
+	 .exp_err = ERANGE,
+	 },
+	{			/* case 03, long value */
```
Looking into fs/xattr.c it looks like we get ERANGE for empty string as a key as well, we should probably cover that one as well.

rafaeldtinoco replied:
Will do.

rafaeldtinoco replied:
Done in fsxattr01 -> case 07

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,202 @@
+
+static void cleanup(void)
+{
+	if (filed >0)
```
Missing space after '>', shouldn't checkpatch.pl complain about these?

rafaeldtinoco replied:
Errr, my bad here. Haven't ran it in the last round before pushing.

rafaeldtinoco commented:
Thanks on reviewing all this! I'm off this week, but will fix everything and repush asap beginning next.

rafaeldtinoco commented:
Hello Cyril, 
I believe I have covered all raised issues, please let me know if not. While checking ERANGE cases, I noticed there was another case: NULL key -> EFAULT, so I added it to fsetxattr01 test as well. Please note that your review was made on fsetxattr{01,02} files only and not in the 2nd commit, the fgetxattr{01,02,03} ones. Since the issues were pretty much the same, I fixed them in those as well.
Thanks for reviewing.
Rafael

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,172 @@
+#define FNAME MNTPOINT"/fgetxattr01testfile"
+
+#define SAFE_FREE(p) { if (p) { free(p); (p) = NULL; } }
+
```
free(NULL) is a no-op, so we can just call free() in the cleanup() instead.

rafaeldtinoco replied:
yep, i saw it in couple of other tests and also thought this could go into lib but was in doubt. will change.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,172 @@
+		}
+
+		tst_res(TFAIL, "fgetxattr(2) passed unexpectedly");
+		goto check_value;
```
It's a bit puzzling that we jump down below just to execute code that is otherwise unreachable.

rafaeldtinoco replied:
will revisit this for the sake of better readability.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,172 @@
+	SAFE_TOUCH(FNAME, 0644, NULL);
+	fd = SAFE_OPEN(FNAME, O_RDONLY, NULL);
+	SAFE_UNLINK(FNAME);
+
```
We actually don't have to unlink the file here since it will be deleted once the test temporary directory is removed.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,172 @@
+	fd = SAFE_OPEN(FNAME, O_RDONLY, NULL);
+	SAFE_UNLINK(FNAME);
+
+	for (i = 0; i < (int) ARRAY_SIZE(tc); i++)
```
If you defined the i to be unsigned or size_t you could have avoided this cast.

rafaeldtinoco replied:
errr... =) things you do in auto mode, tks for catching.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr03.c:
```diff
@@ -0,0 +1,92 @@
+
+static struct tst_test test = {
+	.setup = setup,
+	.test = verify_fgetxattr,
```
If the testcase does only one test you can use the .test_all function pointer instead.

rafaeldtinoco replied:
my bad here, i should have got this from "Test-Writing-Guidelines". will be extra careful in next tests not to add extra burden on your review.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr02.c:
```diff
@@ -0,0 +1,287 @@
+	SAFE_MKNOD(BLK, S_IFBLK | 0777, dev);
+
+	for (i = 0; i < (int) ARRAY_SIZE(tc); i++) {
+
```
Here as well unsigned int or size_t.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr02.c:
```diff
@@ -0,0 +1,287 @@
+
+		tst_res(TFAIL, "fgetxattr(2) on %s passed unexpectedly %ld",
+				tc[i].fname + OFFSET, TEST_RETURN);
+		goto check_value;
```
Here as well a bit puzzling.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr02.c:
```diff
@@ -0,0 +1,287 @@
+	if (tc[i].exp_ret >= 0)
+		SAFE_FSETXATTR(tc[i].fd, tc[i].key, tc[i].value,
+				tc[i].size, tc[i].flags);
+
```
This probably belongs to setup(). I suppose that this will fail the test with EEXISTS if we pass -i 2 parameter to the testcase.

rafaeldtinoco replied:
will fix.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr02.c:
```diff
@@ -0,0 +1,287 @@
+{
+	int i = 0;
+
+	for (i = 0; i < (int) ARRAY_SIZE(tc); i++) {
```
Here as well.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,219 @@
+ * 8. Create new attr whose key length is zero,
+ *    fsetxattr(2) should return -1 and set errno to ERANGE
+ * 9. Create new attr whose key is NULL,
+ *    fsetxattr(2) should return -1 and set errno to EFAULT
```
There are architectures that does have a mapping at NULL address, if I remember correctly it was ia64 and this will also break the test on uClinux. While these are corner cases using fixed value for EFAULT does not seem to be safe enough. We do have tst_get_bad_addr() helper that maps a page with PROT_NONE and returns pointer suitable for EFAULT tests.

rafaeldtinoco replied:
will revisit.

rafaeldtinoco replied:
Done in fsetxattr01 -> case 08.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,250 @@
+
+	if (TEST_RETURN == 0) {
+		tst_res(TFAIL, "fsetxattr() on %s passed unexpectedly",
+				tc[i].fname + OFFSET);
```
We can probably shorten the filenames, since the test temporary directory is unique and the name is based on the test name. That way we don't have to use the offset, but either way it's fine I guess.

rafaeldtinoco replied:
yep, thought it was long also. got it from setxattr()/getxattr() tests. will keep like that for the sake of being equal to other tests only since its fine.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,250 @@
+{
+	int i = 0;
+
+	for (i = 0; i < (int) ARRAY_SIZE(tc); i++)
```
Here as well.

rafaeldtinoco commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,172 @@
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of version 2 of the GNU General Public
+ * License as published by the Free Software Foundation.
+ *
```
will get it from there.

rafaeldtinoco commented:
I re-pushed both commits with all mentioned changes (but the filename extension change, to make it similar to other existing tests). Thank you for reviewing it.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,207 @@
+};
+struct test_case tc[] = {
+	{			/* case 00, invalid flags */
+	 .key = long_key,
```
Shouldn't this be .key = XATTR_TEST_KEY?

Because otherwise this depends on the order of the checks in kernel (which is not stable by any means) and we expect the invalid flag test to kick in first.

rafaeldtinoco replied:
Hum, I brought this from setxattr01.c. May I fix this in both in a separate commit in this merge request ? Only test needing a long_key is the case 02, like you saw. I agree that we depend on ordering on both and it shouldn't be like this.

rafaeldtinoco replied:
Done in fsetxattr01.c and fsetxattr02.c.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,207 @@
+	 .size = 0,
+	 .flags = XATTR_CREATE,
+	 .exp_err = 0,
+	 },
```
There is a problem with this particular test case. As the tests can be passed a -i parameter to execute the testcases several times in a loop this will create the key on the first iteration and which will make the REPLACE test pass unexpectedly in the second iteration and this particular testcase will fail with EEXIST.

I suppose that the easiest solution would be to move the zero length test to a separate testcase and create an existing key for the EEXIST test for XATTR_CREATE in the test setup.

Sorry for not catching that sooner.

metan-ucw replied:
Or we may remove the attribute if the call was successful, we will need to create a separate attribute for the EEXIST testcase in the test setup though...

rafaeldtinoco replied:
Yes, I assume this is only for case 01 (where we want to replace a non-existing attribute, and that won't be true for the 2nd iteration). Fixing it...

rafaeldtinoco replied:
Nah, I'll do what you said on having flags: Check if test demands existing flag and set it before the test. In all other cases, if successful I'll remove it.

rafaeldtinoco replied:
Same as above.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr02.c:
```diff
@@ -0,0 +1,234 @@
+	 .size = XATTR_TEST_VALUE_SIZE,
+	 .flags = XATTR_CREATE,
+	 .exp_err = 0,
+	 },
```
And there is a similar problem for the positive tests here as well, they fail on second iteration with EEXIST. I guess that we may fix that by removing the attribute if the call was successful.

rafaeldtinoco replied:
Wasn't testing iterations before pushing. Will be extra-careful on those from now on to avoid those silly mistakes and extra overhead to you. Thanks.

rafaeldtinoco replied:
Same as the above. Will cleanup key after set and if successful.

metan-ucw commented on testcases/kernel/syscalls/fsetxattr/fsetxattr01.c:
```diff
@@ -0,0 +1,207 @@
+
+	for (i = 0; i < ARRAY_SIZE(tc); i++)
+		if (!tc[i].key)
+			tc[i].key = tst_get_bad_addr(cleanup);
```
Also this is very minor, but I think that LKML  prefers curly braces around block that spans for more than one line.

rafaeldtinoco replied:
Will fix.

metan-ucw commented on testcases/kernel/syscalls/fgetxattr/fgetxattr01.c:
```diff
@@ -0,0 +1,146 @@
+		if (strcmp(tc[i].value, XATTR_TEST_VALUE))
+			tst_res(TFAIL, "wrong value, expect \"%s\" got \"%s\"",
+					 XATTR_TEST_VALUE, tc[i].value);
+
```
Actually this check fails for me because the XATTR_TEST_VALUE_SIZE does not account for the terminating null byte and there is random non-zero byte after the buffer we use for the fgetxattr().

Happily this is easily fixed if we replace the XATTR_TEST_VALUE_SIZE with sizeof(XATTR_TEST_VALUE).

rafaeldtinoco commented on testcases/kernel/syscalls/fgetxattr/fgetxattr03.c:
```diff
@@ -0,0 +1,74 @@
+	.test_all = verify_fgetxattr,
+	.cleanup = cleanup,
+	.needs_tmpdir = 1,
+	.needs_root = 1,
```
Probably not. Will review all of them, likely just the f{g,s}etxattr01 ones.

rafaeldtinoco commented:
I'm fixing the 'attr/xattr.h' issue and submitting again.

metan-ucw commented:
Pushed with two minor changes, good work, thanks!

I've changed the fgetxattr03.c to print TST_RET on failure and removed two /* root: check all supported filesystems */ comments that I think are not that useful, I hope that you do not mind.

Also I guess that we may also change the test library to set the needs_root bit automatically if all_filesystems is set.

rafaeldtinoco commented:
I do not mind at all. Thanks for your review. Next requests, through mailing list, will be more straight forward for you since I'm getting used to the code base much more. Tks again.
